Plotcounts & more...

silibum opened the pull request:
This PR has the main aim to fix the plot size bug that is around a few patches long now. I removed Harmony as it blows up the binary by around 1MB and it isn't needed any more (we talked about that @thomasfn).

Also I created a few wishlist-features and some things I wanted to have overall... Changelog below.

_Tested with a Backup of White-Tiger_

**General**
- Removed HarmonyPatches
- Updated Packages & Modkit
- prestaging workflow -> build on all branches for validation (there is a naming flaw as I left over the old name for the resulting binary-archive as "mod-binaries-staging".. please feel free to ignore this for now...)
- Misc code changes / applied some intellisense suggestions
- Added copy to clipboard to the join mails for accept/reject and make it more outstanding by adding a header
- Do not invite residents back to the HQ (which the game should not do at all ...)
- Configuration is now supporting categories in Server-GUI

**RefreshPlotSize & Calculation at start**
- Now uses OnPostInitialize() from Eco.Server to set correct plotcounts at serverstart
- RefreshPlotsSize() => UpdateClaimData is the new ClaimUpdate-Method
- DirectCitizenship has to be updated earlier to avoid gettin' a hole from the legal person in a city/state/federation

 
**Reputation**
  - Intercept reputation giving to other employess and from/to invited employees (settings -> CompanyReputationInterceptionEnabled | disabled by default)
  - Deny reputation gains on the legal person by other players (no setting, fixed)
  - Reputation on the legal person (if enabled): All employees reputation as average (negative and positive) (settings -> ReputationAveragesEnabled | disabled by default)  
 - The company members can't receive reputation (settings -> DenyCompanyMembersExternalReputationEnabled | disabled by default)
    (if this setting is disabled the reputation relationsship of the legal person will be cleared as 'silent migration' to the new default of 'no reputation possible for the legal person')
- New and changes chatcommands for reputation:
  - /company invites -> list alle active invites for the you or a hint that you are already a member of a company ...
  - /company force -> verb 'join' -> simplified code
  - /company force -> verb 'demote' -> demotes the player to a normal employee (mainly for testing/debugging, but useful)
  - /company force -> verb 'promote' -> now removes the promoted player from the basic employee list as the ceo will prepended to this and a promot leaded to wrong plotcounts as the ceo + 'employee-ceo' was counted.
  - /company configure -> new and adminonly to enabled/disable features on the fly (which are now: 'propertyLimits', 'reputationAverages' or 'reputationInterception')

**Vehicle transfer**
  - on join and on placement (config: VehicleTransfersEnabled | disabled by default)
  - vehicle deeds are now filtered at /company claim due to this
  - takeover is possible
  - deeds get a name for the legal person and the deed ID (as I don't want to count the instance of every vehicle for the legal person...)
 
**New / changed chat commands**
- /company status (renamed from config) -> shows status of config options
- /company configure (admin only) -> sets config options on the fly where valid verbs are 'propertyLimits', 'reputationAverages',  'reputationInterception' or 'vehicleTransfers'
- /company list (admin only) -> shows a list of all companies for information gathering 
- /company rent -> list all deeds that are not the HQ to edit the rents for them (make them renable)
- /company force with verb invite
- /company force with verb uninvite

**Misc**
- Try/ForceJoin and TryFire aren't silent anymore as these actions can be intercepted with laws it makes sense to show a possible rejection by law

**Internal notes**
- RemoveCeo() was added to deduplicate code and does a little check if there is ceo as it also adds the user as employee
- 2 new reputation givers were added and are based on the SpeakingWellOfOthersReputationGiver / unsure if this is the best was, needs more testing.
- IsInvited() as new helper
- Added an automatic takeover of placed claims to the company if PropertyLimitsEnabled is active
- Added a workaround for ECO-36228 which does not clean lifted claimstakes associated deeds and does NOT trigger a PropertManager.DeedDestroyedEvent since V11 (will be removed once the bug is fixed)
- Added a the new ignoreOwnerChanged variable in Company.cs to disable event handling on special ownership changes (like mange the rent -> the dialog will only open if we are the owner, but we are just a microsend the owner to open the UI)

silibum committed:
Basic rework / optimize

- Removed HarmonyPatches
- Updated Packages
- Removed not working calls (mainly refresh for hq size -> will fix this in a different pr)
- prestaging workflow -> build on all branches for validation

- Misc code changes

silibum committed:
RefreshPlotsSize- & Startfixes

- Now uses OnPostInitialize() from Eco.Server to set correct plotcounts at serverstart
- RefreshPlotsSize() => UpdateClaimData is the new ClaimUpdate-Method

silibum committed:
Reputation & more

Reputation
  - Intercept reputation giving to other employess and from/to invited employees (settings -> CompanyReputationInterceptionEnabled | disabled by default)
  - Deny reputation gains on the legal person by other players (no setting, fixed)
  - Reputation on the legal person (if enabled): All employees reputation as average (negative and positive) (settings -> ReputationAveragesEnabled | disabled by default)
    (if this setting is disabled the reputation relationsship of the legal person will be cleared as 'silent migration' to the new default of 'no reputation possible for the legal person')
- New and changes chatcommands:
  - /company invites -> list alle active invites for the you or a hint that you are already a member of a company ...
  - /company force -> verb 'join' -> simplified code
  - /company force -> verb 'demote' -> demotes the player to a normal employee (mainly for testing/debugging, but useful)
  - /company force -> verb 'promote' -> now removes the promoted player from the basic employee list as the ceo will prepended to this and a promot leaded to wrong plotcounts as the ceo + 'employee-ceo' was counted.
  - /company confiugure -> new and adminonly to enabled/disable features on the fly (which are now: 'propertyLimits', 'reputationAverages' or 'reputationInterception')
- Do not invite residents back to the HQ (which the game should not do at all ...)
 
Misc
- Try/ForceJoin and TryFire aren't silent anymore as these actions can be intercepted with laws it makes sense to show a possible rejection by law

Internal notes
- RemoveCeo() was added to deduplicate code and does a little check if there is ceo as it also adds the user as employee
- 2 new reputation givers were added and are based on the SpeakingWellOfOthersReputationGiver / unsure if this is the best was, needs more testing.
- IsInvited() as new helper

silibum committed:
Fix 'holes'

- DirectCitizenship has to be updated earlier to avoid gettin' a hole from the legal person in a city/state/federation

silibum committed:
Vehicle transfers & more

**Vehicle transfer**
  - on join and on placement (config: VehicleTransfersEnabled | disabled by default)
  - vehicle deeds are now filtered at /company claim due to this
  - takeover is possible
  - deeds get a name for the legal person and the deed ID (as I don't want to count the instance of every vehicle for the legal person...)
 
**New and changed chat commands**
- /company status (renamed from config) -> shows status of config options
- /company configure (admin only) -> sets config options on the fly where valid verbs are 'propertyLimits', 'reputationAverages',  'reputationInterception' or 'vehicleTransfers'
- /company list (admin only) -> shows a list of all companies for information gathering

silibum committed:
Fix / Change

- removeCeo is now AddCeoAsEmployee as it does that instead of removeCeo
- fixed force demote to work now (didn't cleared the current ceo but used removeCeo which now does what it is named...)

thomasfn commented on .github/workflows/prestaging.yml:
```diff
@@ -0,0 +1,42 @@
+name: build for prestatging
```
`prestatging` should be `prestaging`

thomasfn commented on EcoCompaniesMod/CompaniesPlugin.cs:
```diff
@@ -4,8 +4,9 @@
 using System.Reflection;
 using System.Collections.Generic;
+using System.Threading.Tasks;
+using Eco.Server;
```
As per convention, this using should be in the scoped using block below as just `using Server;` (and probably in it's own paragraph)

thomasfn commented on EcoCompaniesMod/CompaniesPlugin.cs:
```diff
@@ -44,16 +45,25 @@ public class CompaniesConfig
+        public bool ReputationAveragesEnabled { get; set; } = false;
+
+		[LocDescription("If enabled, the company vehicles will be adopted to the legal person on placement.")]
+		public bool VehicleTransfersEnabled { get; set; } = false;
```
Inconsistent indents, should be using spaces

thomasfn commented on EcoCompaniesMod/CompaniesPlugin.cs:
```diff
@@ -87,6 +97,17 @@ public PostResult Perform(GameAction action, AccountChangeSet acc)
                 case PlaceOrPickUpObject placeOrPickupObjectAction:
                     CompanyManager.Obj.InterceptPlaceOrPickupObjectGameAction(placeOrPickupObjectAction, ref result);
                     break;
+                case ReputationTransfer transferTransferAction: // intercepts new reputation actions
```
`transferTransferAction` variable should be `reputationTransferAction`

thomasfn commented on EcoCompaniesMod/CompaniesPlugin.cs:
```diff
@@ -87,6 +97,17 @@ public PostResult Perform(GameAction action, AccountChangeSet acc)
+                    CompanyManager.Obj.InterceptReputationTransfer(transferTransferAction, ref result);
+                    if (result.Success) // update both sides if we had success
+                    {
+                        Task.Delay(1000).ContinueWith((t) => // wait for ticks
```
Is there any particular science or reasoning to 1000ms of delay? I know I've used 250ms elsewhere to do things "later", I wonder if we should introduce a constant somewhere and make it consistent throughout the codebase

silibum replied:
The long delays where for me to debug. I made it constant to 250 for all tasks now per constant.

thomasfn commented on EcoCompaniesMod/CompaniesPlugin.cs:
```diff
@@ -114,10 +135,7 @@ public class CompaniesPlugin : Singleton<CompaniesPlugin>, IModKitPlugin, IConfi
-            //var harmony = new Harmony("Eco.Mods.Companies");
-            //harmony.PatchAll(Assembly.GetExecutingAssembly());
+
         }
```
May as well get rid of the entire constructor now since it's not needed

thomasfn commented on EcoCompaniesMod/CompaniesPlugin.cs:
```diff
@@ -146,6 +166,11 @@ public void Initialize(TimedTask timer)
+        internal static void OnPostInitialize() {
+            Registrars.Get<Company>().ForEach(company => company.RefreshHQPlotsSize());
+            Logger.Info($"Set correct plot counts for HQs...");
+        }
```
Incorrect bracket placement (should be on new line). Also I don't think the log message is useful beyond debugging, so it could either be removed or changed to debug category

thomasfn commented on EcoCompaniesMod/Company.cs:
```diff
@@ -175,15 +184,12 @@ public override void Initialize()
         {
-            RefreshHQPlotsSize();
+
         }
```
Can be removed entirely

thomasfn commented on EcoCompaniesMod/Company.cs:
```diff
@@ -38,7 +38,8 @@ namespace Eco.Mods.Companies
     using Shared.Utils;
-    using Eco.Core.Plugins;
+    using Eco.Gameplay.Economy.Reputation;
+    using System.Threading.Tasks;
```
As before, usings need tidying, namely the System using goes to the top and the Eco usings trimmed and put into their correct paragraphs (yes I know the using style is weird, I wouldn't do it like this again if I rewrote all the mods from scratch but it is what it is)

thomasfn commented on EcoCompaniesMod/Company.cs:
```diff
@@ -59,7 +60,10 @@ public class Company : SimpleEntry, IHasIcon
+		public static bool IsInvited(User user, Company company)
+			=> company.InviteList.Contains(user);
+
+		public static Company GetEmployer(User user)
```
Incorrect indenting

thomasfn commented on EcoCompaniesMod/Company.cs:
```diff
@@ -139,6 +146,8 @@ public override void Initialize()
 
             // Setup employees
             Employees ??= new ThreadSafeHashSet<User>();
+            PositiveReputationGiver ??= new();
+            NegativeReputationGiver ??= new();
```
Did you verify that `Initialize` is indeed called on every startup as well as when the company is first founded? As I say if it isn't we have some issues to sort out, but namely this would need a migration strategy for old companies that don't have these objects set. If init is called every startup this is a non-issue

silibum replied:
Yes all cool there - what I had as we talked was through a bad migration of a save and solved after restoring the original save back.

thomasfn commented on EcoCompaniesMod/Company.cs:
```diff
@@ -752,15 +759,122 @@ public void UpdateAllAuthLists()
+        {
+            foreach (var sourceUser in AllEmployees)
+            {
+                foreach (var targetUser in AllEmployees.Where(x => x.Name != sourceUser.Name))
```
Is comparing the name necessary? I thought there was a linq verb like `.Exclude` or something to exclude a certain element from the sequence, otherwise just a straight comparison of the `User` object should be fine

silibum replied:
This is now handled as `.Except(sourceUser.SingleItemAsEnumerable())` as `.Except` needs it as Enumerable this works and no more name compare (which indeed was bit to detailed here)

thomasfn commented on EcoCompaniesMod/Company.cs:
```diff
@@ -752,15 +759,122 @@ public void UpdateAllAuthLists()
+            var _reputationObj = ReputationManager.Obj.GetReputation(LegalPerson);
+            _reputationObj?.Relationships.Clear(); // we clear all legal person reputation to remove reputation (silent migration)
+
+            if (!CompaniesPlugin.Obj.Config.ReputationAveragesEnabled) { return; }
```
I think it would be better if the legal person rep block is behind the feature flag too. So if the feature flag is off, players can still rep the legal person as before, and blocking/diverting it only happens if the feature flag is on.

silibum replied:
See new commit - this got a bit revamped / comments on the new variant appreciated :)

thomasfn commented on EcoCompaniesMod/Company.cs:
```diff
@@ -752,15 +759,122 @@ public void UpdateAllAuthLists()
+        public void UpdateLegalPersonReputation()
+        {
+            var _reputationObj = ReputationManager.Obj.GetReputation(LegalPerson);
+            _reputationObj?.Relationships.Clear(); // we clear all legal person reputation to remove reputation (silent migration)
```
Don't need to prefix local variables with an underscore. Also other cases of this in this method

thomasfn commented on EcoCompaniesMod/Company.cs:
```diff
@@ -752,15 +759,122 @@ public void UpdateAllAuthLists()
+
+            CleanSelfReputation(); // after this we have to wait a bit ...
+
+            Task.Delay(2500).ContinueWith((t) =>
```
As before with the science or reasoning about the 2500ms delay, and if it could be made consistent.

thomasfn commented on EcoCompaniesMod/Company.cs:
```diff
@@ -752,15 +759,122 @@ public void UpdateAllAuthLists()
+                    // Logger.Info($"{positiveReputation} + {negativeReputation} = {relativeReputation} | {user.Name}");
+                }
+
+				reputationCountPostive = (reputationCountPostive != 0) ? reputationCountPostive / AllEmployees.Count() : reputationCountPostive;
```
Incorrect indenting

thomasfn commented on EcoCompaniesMod/Company.cs:
```diff
@@ -752,15 +759,122 @@ public void UpdateAllAuthLists()
+                    {
+                        ReputationManager.Obj.GetReputation(targetUser)?.AdjustRelationship(sourceUser, -givenReputation, null, true);
+						// Logger.Debug($"{targetUser.Name} <-> {givenReputation} | {sourceUser.Name}");
+						if (ReputationManager.Obj.ReputationGivenToday(sourceUser, targetUser) != 0)
```
Incorrect indenting

thomasfn commented on EcoCompaniesMod/Company.cs:
```diff
@@ -752,15 +759,122 @@ public void UpdateAllAuthLists()
+
+                    var _sourceName = obj.Creator?.Name ?? obj.Owner?.Name; // WT #43 -> creators where empty on some objects!? migration?
+                    var _newName = obj.Name.Replace(_sourceName, LegalPerson.Name);
+                    var _newNameParts = _newName.Split(' ');
```
As above with underscores on local variables

thomasfn commented on EcoCompaniesMod/Company.cs:
```diff
@@ -896,11 +1008,38 @@ private void MarkPerUserTooltipDirty(User user)
             ServiceHolder<ITooltipSubscriptions>.Obj.MarkTooltipPartDirty(nameof(PerUserTooltip), instance: this, user: user);
         }
 
+        public bool DemoteCeo(User currentCeo)
+        {
+            if (currentCeo != Ceo)
+            {
+                return false;
+            }
+
+            SendCompanyMessage(Localizer.Do($"{currentCeo.UILink()} has been removed as CEO."));
+			AddCeoAsEmployee();
+            Ceo = null;
+
+            return true;
+        }
```
I'm not sure if I like the idea of the possibility that a company can have no CEO. If nothing else it introduces a bunch of possible NREs since everywhere always assumes a CEO exists so far. What's the reasoning behind this change?

silibum replied:
I checked that and could not see any NREs we would create it beside this is more for debugging.
Most calls/uses of the CEO are `!= null` or links that are generated by `UILinkNullSafe()` or when not in the new company creation logic where we will have one...

thomasfn commented on EcoCompaniesMod/Company.cs:
```diff
@@ -937,26 +1076,24 @@ public LocString Tooltip()
             sb.Append(TextLoc.HeaderLoc($"CEO: "));
             sb.AppendLine(Ceo.UILinkNullSafe());
             sb.AppendLine(TextLoc.HeaderLoc($"Employees:"));
-            sb.AppendLine(this.Employees.Any() ? this.Employees.Select(x => x.UILinkNullSafe()).InlineFoldoutListLoc("citizen", TooltipOrigin.None, 5) : Localizer.DoStr("None."));
+            sb.AppendLine(Employees.Any() ? Employees.Select(x => x.UILinkNullSafe()).InlineFoldoutListLoc("citizen", TooltipOrigin.None, 5) : Localizer.DoStr("None."));
             sb.Append(TextLoc.HeaderLoc($"Finances: "));
-            sb.AppendLine(this.OwnedAccounts.Any() ? this.OwnedAccounts.Select(x => x.UILinkNullSafe()).InlineFoldoutListLoc("account", TooltipOrigin.None, 5) : Localizer.DoStr("None."));
+            sb.AppendLine(OwnedAccounts.Any() ? OwnedAccounts.Select(x => x.UILinkNullSafe()).InlineFoldoutListLoc("account", TooltipOrigin.None, 5) : Localizer.DoStr("None."));
             sb.Append(TextLoc.HeaderLoc($"HQ: "));
-            sb.AppendLine(this.HQDeed != null ? this.HQDeed.UILink() : Localizer.DoStr("None."));
+            sb.AppendLine(HQDeed != null ? HQDeed.UILink() : Localizer.DoStr("None."));
             sb.AppendLine(TextLoc.HeaderLoc($"Property:"));
-            sb.AppendLine(this.OwnedDeeds.Any() ? this.OwnedDeeds.Where(x => x != HQDeed).Select(x => x.UILinkNullSafe()).InlineFoldoutListLoc("deed", TooltipOrigin.None, 5) : Localizer.DoStr("None."));
+            sb.AppendLine(OwnedDeeds.Any() ? OwnedDeeds.Where(x => x != HQDeed).Select(x => x.UILinkNullSafe()).InlineFoldoutListLoc("deed", TooltipOrigin.None, 5) : Localizer.DoStr("None."));
             sb.AppendLine(TextLoc.HeaderLoc($"Shareholders:"));
-            sb.AppendLine(this.Shareholders.Any() ? this.Shareholders.Select(x => x.Description).InlineFoldoutListLoc("holding", TooltipOrigin.None, 5) : Localizer.DoStr("None."));
+            sb.AppendLine(Shareholders.Any() ? Shareholders.Select(x => x.Description).InlineFoldoutListLoc("holding", TooltipOrigin.None, 5) : Localizer.DoStr("None."));
             sb.Append(TextLoc.HeaderLoc($"Citizenship: "));
-            sb.AppendLine(this.DirectCitizenship != null ? DirectCitizenship.UILink() : Localizer.DoStr("None."));
+            sb.AppendLine(DirectCitizenship != null ? DirectCitizenship.UILink() : Localizer.DoStr("None."));
             return sb.ToLocString();
         }
 
         [NewTooltip(CacheAs.Instance | CacheAs.User, 110)]
         public LocString PerUserTooltip(User user)
         {
-            var sb = new LocStringBuilder();
-            if (user == Ceo)
-            {
+            if (user == Ceo) {
```
Incorrect bracketing

thomasfn commented on EcoCompaniesMod/CompanyCommands.cs:
```diff
@@ -1,8 +1,11 @@
 ﻿using System.Threading.Tasks;
 using System.Linq;
+using Eco.Gameplay.Utils;
 
 namespace Eco.Mods.Companies
 {
+	using Eco.Core.Systems;
+
     using Shared.Localization;
     using Shared.Utils;
 
```
As before with using tidy up

thomasfn commented on EcoCompaniesMod/CompanyCommands.cs:
```diff
@@ -14,24 +17,32 @@ namespace Eco.Mods.Companies
     public static class CompanyCommands
     {
         [ChatCommand("Company", ChatAuthorizationLevel.User)]
         public static void Company() { }
 
-        [ChatSubCommand("Company", "Check company mod configuration.", ChatAuthorizationLevel.User)]
-        public static void Config(User user)
+		[ChatSubCommand("Company", "Check company mod configuration.", ChatAuthorizationLevel.User)]
+        public static void Status(User user)
```
The indenting in this file is a disaster, I won't comment on each case but it all needs normalising to use spaces. I'm not even sure I can properly read and review this diff in this state.

thomasfn commented on EcoCompaniesMod/EcoCompaniesMod.csproj:
```diff
@@ -35,18 +35,22 @@
   </ItemGroup>
 
   <ItemGroup>
     <PackageReference Include="Costura.Fody" Version="5.7.0">
       <PrivateAssets>all</PrivateAssets>
     </PackageReference>
-    <PackageReference Include="Fody" Version="6.6.2">
+    <PackageReference Include="Fody" Version="6.8.1">
       <PrivateAssets>all</PrivateAssets>
     </PackageReference>
-    <PackageReference Include="Lib.Harmony" Version="2.2.1" />
+    <PackageReference Include="Mono.Cecil" Version="0.11.5" />
+    <PackageReference Include="MonoMod.Common" Version="22.6.3.1" />
     <PackageReference Include="NetFabric.Hyperlinq.Abstractions" Version="1.3.0" />
-    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
+    <PackageReference Include="Newtonsoft.Json" Version="13.0.3" />
```
`Fody` and `Constura` are only being used to embed the harmony assembly in the mod dll, since harmony is removed, they can go too. Also `Mono.Cecil` and `MonoMod.Common` are dependencies of harmony and can go also.

thomasfn commented on EcoCompaniesMod/Reputation/CompanyReputationGivers.cs:
```diff
@@ -0,0 +1,56 @@
+﻿
```
Remove leading empty line

thomasfn commented on FodyWeavers.xml:
```diff
@@ -1,7 +1,6 @@
 <Weavers xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="FodyWeavers.xsd">
 	<Costura>
 		<IncludeAssemblies>
-			0Harmony
 			MonoMod.Common
 			Mono.Cecil
 		</IncludeAssemblies>
```
As above `Fody` is only used to embed harmony so this whole file can go away now too

thomasfn commented on EcoCompaniesMod/Reputation/CompanyReputationGivers.cs:
```diff
@@ -0,0 +1,56 @@
+	[Serialized]
+	public class CompanyPositiveReputationGiver : IGivesReputation
+	{		
+		public LocString MarkedUpName => TextLoc.InfoLightLocStr("AVG Positive Reputation");
```
I think it'd be better to just say `"Average Positive Reputation"` in the name here, shortening it is ok to help code clarity but it's better to be clear with user-facing text imo.

silibum replied:
![image](https://github.com/user-attachments/assets/f720367a-ebfe-49ff-9f70-f2345af80b96)

Played around on that a bit and changed to a more colored version including "Employee" instead of "Positive/Negative" to make it more clear and visual.

thomasfn commented on EcoCompaniesMod/Reputation/CompanyReputationGivers.cs:
```diff
@@ -0,0 +1,56 @@
+		int controllerID;
+		public ref int ControllerID => ref controllerID;
+
+#pragma warning disable CS0067
```
Which warning is this?

silibum replied:
Outdated, I removed PropertyChanged handling entirely as it isn't needed for this "clone" of the SpeakingWellOfOthersReputationGiver

thomasfn commented on EcoCompaniesMod/Reputation/CompanyReputationGivers.cs:
```diff
@@ -0,0 +1,56 @@
+﻿
+using Eco.Core.PropertyHandling;
+using Eco.Core.Controller;
+
+namespace Eco.Gameplay.Economy.Reputation
+{
+	using System.ComponentModel;
+	using Eco.Shared.Localization;
+	using Eco.Shared.Serialization;
+	using Eco.Shared.Utils;
+	using Range = Eco.Shared.Math.Range;
```
Using tidy up as before (System usings to head of file, Eco usings to scoped paragraphs with prefixes trimmed).

thomasfn commented on EcoCompaniesMod/CompanyManager.cs:
```diff
@@ -252,6 +252,59 @@ public LazyResult ShouldOverrideAuth(IAlias alias, IOwned property, GameAction a
                     return LazyResult.FailedNoMessage;
             }
         }
+		public void InterceptReputationTransfer(ReputationTransfer reputationTransferData, ref PostResult lawPostResult)
```
Indents need fixing as before

silibum committed:
Cleanup / Tidy / Fixes

Clean / Tidy
- untabified code (avoided to plain new files that would create a full changed file now...)
- staging release-686 as modkit branch/version
- PR-36#1713816528 -> (and other comments) cleaned usings
- PR-36#1713789914 -> removed prestaging as I will do another approach für my development
- PR-36#1713802152 -> (and other comments) corrected bracket placing(s)
- PR-36#1713796625 -> reputationTransferAction instead of transferTransferAction
- PR-36#1713799216 -> `static CompaniesPlugin()` removed
- PR-36#1713813890 -> `private void OnDeserialized()` removed
- PR-36#1713798062 -> unified task delay after some testing and tuning
- PR-36#1713830115 -> removed local variable prefixes
- PR-36#1713859875 -> removed fody & co

Reputation (PR-36#1713829465, PR-36#1713827515, PR-36#1713862682)
- added ReputationAveragesBonusEnabled as new option to include or exclude bonus reputation givers
- removed the silent migration and added new options: DenyLegalPersonReputationEnabled & DenyCompanyMembersReputationEnabled so admin can choose how the new feature will work
  (default is now as before the new system)
- simplified the CleanSelfReputation check using `.Exclude()` for the sub-loop
- "Average Positive Reputatio" vs "AVG Positive Reputation" -> Average now and colored positive / negative (light colors)

Vehicle Transfer
- VehicleTransfersUseCompanyNameEnabled as new Option to let the admin decide which way the deeds will be renamed ("Legal Person" as suffix can be looooong)

General
- remade both /company configure and /company status to work dynamic through the settings
- Checked on the "no" CEO thinking: No obvious NREs found as most calls are `!= null` or links that are generated by `UILinkNullSafe()`
- configuration is now supporting categories in Server-GUI

silibum committed:
// Cleanup

- tidy missed a bracket

silibum committed:
// Cleanup

- another formatting fixed

silibum committed:
// Cleanup

- another formatting fixed

silibum committed:
Tuning

- force -> demote now is DevTier flagged internaly (to avoid possible but not expected NREs)
- added copy to clipboard to the join mails for accept/reject and make it more outstanding by adding a header
- new command: /company force with verb invite
- new command: /company force with verb uninvite

silibum committed:
DenyCompanyMembersExternalReputationEnabled

- If enabled, the company members can't receive reputation

(even more flexibel in configuration)

silibum committed:
Tuning & Fixes

- Updated the description for VerhicleTransfersEnabled as it needs PropertyLimitsEnabled also
- Added a command to manage the rent for company owned deeds
- Added an automatic takeover of placed claims to the company if PropertyLimitsEnabled is active as we can manage the rent with a command (take back is still possible)
- Added a workaround for ECO-36228 which does not clean lifted claimstakes associated deeds and does NOT trigger a PropertManager.DeedDestroyedEvent since V11 (will be removed once the bug is fixed)
- Added a the new ignoreOwnerChanged variable in Company.cs to disable event handling on special ownership changes (like mange the rent -> the dialog will only open if we are the owner, but we are just a microsend the owner to open the UI)
- Updated workflows

silibum committed:
Typos fixed

silibum committed:
Tuning

thomasfn commented on EcoCompaniesMod/CompaniesPlugin.cs:
```diff
@@ -42,18 +44,39 @@ namespace Eco.Mods.Companies
+		[LocDescription("If enabled, the company members can't receive reputation."), Category("Reputation")]
+		public bool DenyCompanyMembersExternalReputationEnabled { get; set; } = false;
+
+		[LocDescription("If enabled, the company members can't give reputation to each other nor the legal person (also counts for invited members)."), Category("Reputation")]
```
Indenting

thomasfn commented on EcoCompaniesMod/CompaniesPlugin.cs:
```diff
@@ -42,18 +44,39 @@ namespace Eco.Mods.Companies
+		[LocDescription("If enabled, the company members can't give reputation to each other nor the legal person (also counts for invited members)."), Category("Reputation")]
+        public bool DenyCompanyMembersReputationEnabled { get; set; } = false;
+
+        [LocDescription("If enabled, the average repuation from all employees will be given to the legal person (in addition to their own reputation if they have any)."), Category("Reputation")]
+        public bool ReputationAveragesEnabled { get; set; } = false;
+
+        [LocDescription("If enabled, the average repuation from all employees will be filtered by known bonussources (currently only SpeaksWellOfOthersBonus)."), Category("Reputation")]
+        public bool ReputationAveragesBonusEnabled { get; set; } = true;
```
Are the default reputation settings set in a way that is best for WT?

silibum committed:
Typo

- reject command had 'join' as text.

thomasfn commented on EcoCompaniesMod/CompaniesPlugin.cs:
```diff
@@ -100,26 +134,19 @@ public PostResult Perform(GameAction action, AccountChangeSet acc)
     public class CompaniesPlugin : Singleton<CompaniesPlugin>, IModKitPlugin, IConfigurablePlugin, IInitializablePlugin, ISaveablePlugin, IContainsRegistrars
     {
         private bool ignoreBankAccountPermissionsChanged = false;
+        public const int TaskDelay = 250;
 
         public IPluginConfig PluginConfig => config;
 
         private PluginConfig<CompaniesConfig> config;
         public CompaniesConfig Config => config.Config;
 
-        private static readonly Dictionary<Type, GameValueType> gameValueTypeCache = new Dictionary<Type, GameValueType>();
+        private static readonly Dictionary<Type, GameValueType> gameValueTypeCache = new();
```
I assume we can't initialise with `[]` as that's .net 8?

silibum replied:
I just applied [IDE0090 ](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0090) - made sense to me to shorten this.

thomasfn replied:
Yeah don't worry about it, I'm just already used to doing `Dictionary<X, Y> z = []` but that's a .net 8 thing so we can't use it here

thomasfn commented on EcoCompaniesMod/CompanyCommands.cs:
```diff
@@ -22,16 +25,19 @@ public static class CompanyCommands
+        public static void Status(User user)
         {
-            if (CompaniesPlugin.Obj.Config.PropertyLimitsEnabled)
-            {
-                user.MsgLoc($"Company property limits mode is ENABLED.");
-            }
-            else
+            var sb = new LocStringBuilder().AppendLine();
+            foreach (var configOption in CompaniesPlugin.Obj.PluginConfig.ConfigProperties.ToList())
             {
-                user.MsgLoc($"Company property limits mode is DISABLED.");
+                var statusText = CompaniesPlugin.Obj.Config.GetStringPropertyByName(configOption.Key) == "True" ? "enabled" : "disabled";
+                var statusColor = (statusText == "enabled") ? Color.Green : Color.Red;
+
+                sb.AppendLineLoc($"{Text.Color(Color.BlueGrey, configOption.Key)} are {Text.Color(statusColor, statusText)}");
+                sb.AppendLineLoc($"{configOption.Value.Description}\n");
             }
+
+            user.Player?.OpenInfoPanel($"Plugin-Status", sb.ToString(), "pluginSettingsInfo");
```
Is it possible to get a screenshot of how this looks?

silibum replied:
Sure: ![image](https://github.com/user-attachments/assets/e1e05c6d-004d-4a93-80be-7a5a9fa39d7e)

thomasfn replied:
I think I'd rather the title of the UI be "Companies Mod Status" rather than "Plugin-Status" since it's not exactly clear to which plugin it refers at a glance

thomasfn commented on EcoCompaniesMod/CompanyCommands.cs:
```diff
@@ -89,6 +95,51 @@ public static void Uninvite(User user, User otherUser)
+            {
+                targetCompany.SendCompanyMessage(Localizer.Do($"{user.UILink()} has declined the invitation to join the company.")); 
+                user.OkBoxLoc($"You rejected the invitation to {targetCompany.UILink()}");
+            } else
```
`else` should be on a new line from the `}`

thomasfn commented on EcoCompaniesMod/CompanyCommands.cs:
```diff
@@ -89,6 +95,51 @@ public static void Uninvite(User user, User otherUser)
+        {
+            var company = Companies.Company.GetEmployer(user);
+            if (company != null) {
+                user.OkBoxLoc($"You are already bounded to {company.UILink()}...");
```
Would prefer "You are an employee of X." Assuming that someone who is an employee can't be invited and joining a company will cancel all other invites?

silibum replied:
I don't understand the assumption but I try: This message is if some tries to list his invites and is already employed.
You can be invited to as many companies as you like - at the moment you accept any invite you can't list them anymore as you are now employed. That hasn't changed just the list is new - maybe a good addition to clean all invitelist but this can go in a later code iteration.

Texts is changed.

thomasfn commented on EcoCompaniesMod/CompanyCommands.cs:
```diff
@@ -89,6 +95,51 @@ public static void Uninvite(User user, User otherUser)
+                {
+                    if(!sb.ToString().IsSet())
+                    {
+                        sb.AppendLineLoc($"Your are invite to these companies:\n\n");
```
Would prefer "You have invites from the following companies:"

thomasfn commented on EcoCompaniesMod/CompanyCommands.cs:
```diff
@@ -89,6 +95,51 @@ public static void Uninvite(User user, User otherUser)
+
+            if (!sb.ToString().IsSet())
+            {
+                sb.AppendLineLoc($"Your aren't invited any company.") ;
```
Would prefer "You have no pending invites."

silibum committed:
Bugfix

- update auth for taken vehicledeeds directly instead of at the next company action...

thomasfn commented on EcoCompaniesMod/CompanyCommands.cs:
```diff
@@ -302,13 +421,42 @@ public static void Force(User user, Company targetCompany, User targetUser, stri
+                return;
+            }
+
+            user.Msg(sb.ToLocString());
```
Would this be better to use the transaction log ui instead? e.g. the one you get when you open your bank account's transaction history or use `/tax card`

silibum replied:
Chat stays in backlog while the dialog closes by any UI-Click outside of it -> I think it's better to go with the chat here.

thomasfn commented on EcoCompaniesMod/CompanyManager.cs:
```diff
@@ -252,6 +252,76 @@ public LazyResult ShouldOverrideAuth(IAlias alias, IOwned property, GameAction a
+                }
+            }
+
+			var receiverEmployeer = Company.GetEmployer(reputationTransferData.ReputationReceiver);
+			if (CompaniesPlugin.Obj.Config.DenyCompanyMembersExternalReputationEnabled && receiverEmployeer != null)
+            {
+				lawPostResult.Success = false;
+                NotificationManager.ServerMessageToPlayer(
+                    Localizer.Do($"You can't give reputation to a member of a company."),
+                    reputationTransferData.ReputationSender,
+                    NotificationCategory.Reputation,
+                    NotificationStyle.InfoBox
+                );
+
+				return;
+			}
```
indenting

thomasfn commented on EcoCompaniesMod/CompanyManager.cs:
```diff
@@ -252,6 +252,76 @@ public LazyResult ShouldOverrideAuth(IAlias alias, IOwned property, GameAction a
+
+            if (CompaniesPlugin.Obj.Config.DenyCompanyMembersReputationEnabled) // we do not allow the employees to reputate internal, if the settings match
+            {
+                var senderCompany     = Company.GetEmployer(reputationTransferData.ReputationSender);
```
Space between variable and assignment

thomasfn commented on EcoCompaniesMod/CompanyManager.cs:
```diff
@@ -287,37 +357,67 @@ public void InterceptPlaceOrPickupObjectGameAction(PlaceOrPickUpObject placeOrPi
             if (placeOrPickUpObject.PlacedOrPickedUp == PlacedOrPickedUp.PickingUpObject)
             {
+				// workaround V11 bug ECO-36228 which let's empty deeds behind... | save the deed before it's gone
+				var deed = PropertyManager.GetDeedWorldPos(placeOrPickUpObject.ActionLocation.XZ);
+
                 lawPostResult.AddPostEffect(() =>
                 {
-                    Task.Delay(250).ContinueWith(t => FixupHomesteadClaimItems(placeOrPickUpObject.Citizen));
+                    Task.Delay(CompaniesPlugin.TaskDelay).ContinueWith(t => FixupHomesteadClaimItems(placeOrPickUpObject.Citizen));
+
+					// workaround V11 bug ECO-36228 which let's empty deeds behind... | remove the deed after a short delay to let the game catch up
+					if (deed != null && (placeOrPickUpObject.ItemUsed is SettlementClaimStakeItem settlementClaimStake || placeOrPickUpObject.ItemUsed is HomesteadClaimStakeItem homeClaimStake))
+                    {
+                        Task.Delay(CompaniesPlugin.TaskDelay).ContinueWith(t =>
+                        {
+                            try
+                            {
+                                if (deed.OwnedObjects?.Any() != true)
+								{
+                                    // Logger.Debug($"Fixed up empty deed '{deed.Id}' left due to ECO-36228");
+                                    deed.ForceChangeOwners(placeOrPickUpObject.Citizen, OwnerChangeType.CivicUpdate);
+									Registrars.Get<Deed>().Remove(deed);
+                                }
```
indenting

thomasfn commented on EcoCompaniesMod/CompanyManager.cs:
```diff
@@ -287,37 +357,67 @@ public void InterceptPlaceOrPickupObjectGameAction(PlaceOrPickUpObject placeOrPi
+                {
+                    lawPostResult.AddPostEffect(() =>
+                    {
+                        Task.Delay(CompaniesPlugin.TaskDelay).ContinueWith(t => company.UpdateAllVehicles()); // take over vehicle if we got some new
+                        Task.Delay(CompaniesPlugin.TaskDelay).ContinueWith(t => company.TakeClaim(placeOrPickUpObject.Citizen, placeOrPickUpObject.ActionLocation.XZ)); // take claimstake over if it is one (special handling in compare to vehicle)
```
Is there any particular reason there are two delayed tasks here rather than a single one that calls both?

silibum replied:
Yes I decided to seperate them as I don't want all actions to fail if something in that new operation is causing a unexpected NRE or something else. There is also a UpdateAllAuthLists() missing which I just added.

thomasfn commented on EcoCompaniesMod/Reputation/CompanyReputationGivers.cs:
```diff
@@ -0,0 +1,43 @@
+	using Shared.Serialization;
+	using Shared.Utils;
+	using Range = Shared.Math.Range;
+
+	[Serialized]
+	public class CompanyPositiveReputationGiver : IGivesReputation
+	{		
+		public LocString MarkedUpName => TextLoc.InfoLightLocStr(Text.Color(Color.LightGreen, "Average Employee Reputation"));
+
+		[Serialized] public int Id { get; set; } = RandomUtil.IntValue;
+
+		float IGivesReputation.GivableReputationPerDay => float.PositiveInfinity;
+		public float GivableReputationPerDayPerTarget => float.PositiveInfinity;
+		public Range GivableReputationToSingleTargetTotal => new(0, float.PositiveInfinity);
+		int IGivesReputation.DisplayPriority => 2;
+
+		#region IController
+		int controllerID;
+		public ref int ControllerID => ref controllerID;
+		#endregion
+	}
+
+	[Serialized]
+	public class CompanyNegativeReputationGiver : IGivesReputation
+	{
+		public LocString MarkedUpName => TextLoc.InfoLightLocStr(Text.Color(Color.LightRed, "Average Employee Reputation"));
+
+		[Serialized] public int Id { get; set; } = RandomUtil.IntValue;
+
+		float IGivesReputation.GivableReputationPerDay => float.PositiveInfinity;
+		public float GivableReputationPerDayPerTarget => float.PositiveInfinity;
+		public Range GivableReputationToSingleTargetTotal => new(0, float.PositiveInfinity);
+		int IGivesReputation.DisplayPriority => 2;
+
+		#region IController
+		int controllerID;
+		public ref int ControllerID => ref controllerID;
+		#endregion
+	}
+}
```
Indents in this file need changing to spaces

silibum committed:
Indents & Text

thomasfn commented on EcoCompaniesMod/Company.cs:
```diff
@@ -211,7 +218,13 @@ public bool TryInvite(User invoker, User target, out LocString errorMessage)
+            var rejectLink = Text.CopyToClipBoard(Text.Color(Color.Red, $"To reject use"), $"/company reject {Name}", $"/company join {Name}");
+            var headerText = Text.Header($"You have been invited to join {this.UILink()}");
+
+			target.MailLoc($"{headerText}\n\n{acceptLink}\n{rejectLink}", NotificationCategory.Government);
```
indent

thomasfn commented on EcoCompaniesMod/Company.cs:
```diff
@@ -571,8 +586,22 @@ public void RefreshHQPlotsSize()
 
+        public void EditRent(Deed deed, User user)
+        {
+            ignoreOwnerChanged = true; // Don't go through company events for this - we will give it right back after we open the UI.
+
+            var currentOwner = deed.Owner;
+            deed.ForceChangeOwners(user, OwnerChangeType.AdminCommand);
+            deed.Rent.EditRent(user.Player);
+            deed.ForceChangeOwners(currentOwner, OwnerChangeType.AdminCommand);
+
+            ignoreOwnerChanged = false;  // Okay, we're done.
```
Could this be wrapped in a try finally block, on the off-chance that something in there throws? e.g.

```C#
try
{
  ignoreOwnerChanged = true;
  // do stuff
}
finally
{
  ignoreOwnerChanged = false;
}
```

silibum replied:
Totaly like that. Never thought that way - nice hint!

silibum committed:
Indents & Text

silibum committed:
Typo

- reject command had 'join' as text 2nd found.
